core/src/cache_block_meta_service.rs

```diff
@@ -34,23 +38,27 @@ impl CacheBlockMetaService {
                 if exit.load(Ordering::Relaxed) {
                     break;
                 }
-                let recv_result = cache_block_meta_receiver.recv_timeout(Duration::from_secs(1));
-                match recv_result {
-                    Err(RecvTimeoutError::Disconnected) => {
-                        break;
+
+                let bank = match cache_block_meta_receiver.try_recv() {
+                    Ok(bank) => bank,
+                    Err(TryRecvError::Empty) => {
+                        sleep(TRY_RECV_INTERVAL);
```

Comment:
This does introduce a behavior change that could delay updating block height/time for up to 100ms compared to the current behavior.

My gut tells me it's probably no big deal to have this information be 100ms stale for a given RPC query, but I want to at least call this out

Reply:
Note that `getBlockHeight` is retrieved from a Bank and not the Blockstore.  `getBlockTIme` also hits a Bank unless the block is is rooted, so `>= 32 * 400ms` old.

But, these values are both included in `getBlock` and can be `None`:
https://github.com/anza-xyz/agave/blob/34069c481417dfb58d2510bdd9873058961bd9ca/ledger/src/blockstore.rs#L2831-L2835

Some options:
- Keep 100ms sleep and be ok with these values being `None` in `getBlock` 
- We could drop the sleep value, altho I'm not positive what the right value is. 10 ms ? 1 ms ?
    - With 1ms sleep, it would suck to wake up 1000 times a second for the 2.5 events that will come across
- We could do a more sophisticated sleep pattern. For example, we could sleep for 200ms when we see a message come across, and then cut the sleep in half each iteration ( divide-by-2 stops at `1ms`)
- I was against it in the other PR, but maybe we'd be ok with `recv()` (including the spin-loop and yields)
- Use a lower level primitive (like `CondVar`) directly to avoid the `recv()` spins and yields

Reply:
And two other general comments:
- Whatever value we decide on, we would want the same for `RewardsRecorderService` too (https://github.com/anza-xyz/agave/pull/3979). We essentially do the same thing for rewards in `getBlock` that we do if block time and block height are unavailable:
https://github.com/anza-xyz/agave/blob/34069c481417dfb58d2510bdd9873058961bd9ca/ledger/src/blockstore.rs#L2825-L2829
- Instead of repeating logic to get above, we can and should combine `RewardsRecorderService` and `CacheBlockMetaService`; the `Rewards` come from a `Bank` so we can easily shift the work `RewardsRecorderService` does into `CacheBlockMetaService` and reduce our thread count by 1

Reply:
Going to leave this PR open for now for potential discussion, but going to proceed with cleanup of this service (including combining `RewardsRecorderService` into it) first so this PR will likely get closed out